Remove direct dependency on libyang

bhouse-nexthop opened the pull request:
#### What I did

sonic-utilities should never call directly into libyang.  All functionality it needs should be provided in sonic-yang-mgmt.  This resolves issues when libyang changes API/ABI such as happened between libyang1 and libyang2, and again between libyang2 and libyang3.

Depends on https://github.com/sonic-net/sonic-buildimage/pull/24414 and https://github.com/sonic-net/sonic-buildimage/pull/26563
Fixes https://github.com/sonic-net/sonic-buildimage/pull/22385

#### How I did it

Removed calls to libyang and instead uses new functionality introduced in sonic-yang-mgmt.

Also installs both libyang1 and libyang3 for tests to maintain compatibility with both during the migration.

#### How to verify it

After https://github.com/sonic-net/sonic-buildimage/pull/24414 merges observe test cases still pass.

#### Previous command output (if the output of a command-line utility has changed)
N/A

#### New command output (if the output of a command-line utility has changed)
N/A

bhouse-nexthop committed:
Remove direct dependency on libyang

sonic-utilities should never call directly into libyang.  All
functionality it needs should be provided in sonic-yang-mgmt.  This
resolves issues when libyang changes API/ABI such as happened
between libyang1 and libyang2, and again between libyang2 and libyang3.

Depends on https://github.com/sonic-net/sonic-buildimage/pull/24414

Signed-off-by: Brad House <bhouse@nexthop.ai>

saiarcot895 commented on config/config_mgmt.py:
```diff
@@ -71,7 +68,7 @@ def __init__(self, source="configDB", debug=False, allowTablesWithoutYang=True,
         return
 
     def __init_sonic_yang(self):
-        self.sy = sonic_yang.SonicYang(YANG_DIR, debug=self.DEBUG, sonic_yang_options=self.sonicYangOptions)
```
Why can't options be used? Because the underlying enums/constants could change across libyang versions?

bhouse-nexthop replied:
correct, the point is to remove anything that is libyang-specific.  The options here were meaningless anyhow.

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

bhouse-nexthop commented:
/azpw run

mssonicbld commented:
⚠️ **Notice**: `/azpw run` only runs failed jobs now. If you want to trigger a whole pipline run, please rebase your branch or close and reopen the PR.
💡 **Tip**: You can also use `/azpw retry` to retry failed jobs directly.

Retrying failed(or canceled) jobs...

mssonicbld commented:
/azp run

bhouse-nexthop commented:
@saiarcot895 @qiluo-msft this is the next libyang3 blocker

saiarcot895 commented on sonic_package_manager/manager.py:
```diff
@@ -1344,7 +1343,7 @@ def get_manager() -> 'PackageManager':
         docker_api = DockerApi(docker.from_env(), ProgressManager())
         registry_resolver = RegistryResolver()
         metadata_resolver = MetadataResolver(docker_api, registry_resolver)
-        cfg_mgmt = config_mgmt.ConfigMgmt(source=INIT_CFG_JSON, sonicYangOptions=ly.LY_CTX_DISABLE_SEARCHDIR_CWD)
```
Wouldn't this option mean that the current working directory can't be used to resolve any paths in the config being loaded, that it would have to (I guess) all be from the root directory?

bhouse-nexthop replied:
The existing code disabled searching from CWD.

In the libyang3 port of sonic-yang-mgmt we never search CWD for yang models as the python bindings disable searching by default: https://github.com/CESNET/libyang-python/blob/a69eb4a6865069de1a8b73d76b2d7427893214cf/libyang/context.py#L201

And we don't override it when we initialize the context:
https://github.com/bhouse-nexthop/sonic-buildimage/blob/4390075149ae6eb60c2430e74b416f83c33d2e94/src/sonic-yang-mgmt/sonic_yang.py#L56

So the end result is identical.

rookie-who commented:
@bhouse-nexthop Two things:

1. The `find_ref_paths()` change is behavioral — old code walked leaf xpaths in Python and called `find_data_dependencies()` per leaf, new code passes the parent xpath and relies on sonic-yang-mgmt to recurse. Is there a test covering this equivalence? If not, worth adding one to catch any divergence.

2. What prevents a future PR from re-introducing `import yang as ly`? A lint rule or CI check (e.g. grep for `import yang` / `from yang`) would make this a hard gate rather than a convention.

rookie-who commented:
Re: preventing future `import yang` — a semgrep rule would slot right into the existing CI:

```yaml
# .semgrep/no-direct-libyang.yml
rules:
  - id: no-direct-libyang-import
    patterns:
      - pattern-either:
          - pattern: import yang
          - pattern: import yang as $X
          - pattern: from yang import $X
    message: "Do not import libyang directly. Use sonic_yang / sonic_yang_ext instead."
    languages: [python]
    severity: ERROR
    paths:
      exclude:
        - tests/
```

This would fail CI on any PR that reintroduces the dependency. Could be added in this PR or as a follow-up.

rookie-who commented:
Correction on my earlier comment — the existing `find_ref_paths` tests in `gu_common_test.py` do cover non-leaf inputs (`path="/PORT"`, `path=""`), which exercises the changed code path. These should catch any divergence between the old Python leaf-iteration and the new sonic-yang-mgmt recursive call.

The `load_module_str_name()` path in `get_module_name()` is still untested though — worth a quick unit test to cover that.

bhouse-nexthop commented:
@rookie-who

> @bhouse-nexthop Two things:
> 
> 1. The `find_ref_paths()` change is behavioral — old code walked leaf xpaths in Python and called `find_data_dependencies()` per leaf, new code passes the parent xpath and relies on sonic-yang-mgmt to recurse. Is there a test covering this equivalence? If not, worth adding one to catch any divergence.

Yes, the tests most definitely cover this.  Both in the pre-existing sonic-utilities tests as well as tests specifically added to sonic-yang-mgmt for this explicit case.

One of the sonic-yang-mgmt tests that covers this exact scenario is:
https://github.com/sonic-net/sonic-buildimage/blob/406302a9a060a7a40fa05ecbbc2b8c2e4e805943/src/sonic-yang-mgmt/tests/libyang-python-tests/test_SonicYang.json#L68-L76

> 2. What prevents a future PR from re-introducing `import yang as ly`? A lint rule or CI check (e.g. grep for `import yang` / `from yang`) would make this a hard gate rather than a convention.

That's a good point, I can look into how to specifically prevent that.

bradh352 commented:
@rookie-who
> Correction on my earlier comment — the existing `find_ref_paths` tests in `gu_common_test.py` do cover non-leaf inputs (`path="/PORT"`, `path=""`), which exercises the changed code path. These should catch any divergence between the old Python leaf-iteration and the new sonic-yang-mgmt recursive call.
> 
> The `load_module_str_name()` path in `get_module_name()` is still untested though — worth a quick unit test to cover that.

That's covered here: https://github.com/bhouse-nexthop/sonic-utilities/blob/86f1c8ab03eef19531ceca43a6943dbe29b785bd/tests/config_mgmt_test.py#L30

bhouse-nexthop committed:
Add semgrep rule blocking direct libyang imports

Prevent future regressions from re-introducing a direct dependency on
libyang.  The rule covers both libyang1/2 (Python module name 'yang')
and libyang3 (Python module name 'libyang') so neither flavor can be
imported outside of sonic_yang / sonic_yang_ext.

Wires .semgrep/ into the existing Semgrep workflow's SEMGREP_RULES so
the rule runs as a hard gate on PRs and master pushes.

Signed-off-by: Brad House <bhouse@nexthop.ai>

mssonicbld commented:
/azp run

bradh352 commented:
@rookie-who
> Re: preventing future `import yang` — a semgrep rule would slot right into the existing CI:
> 
> ```yaml
> # .semgrep/no-direct-libyang.yml
> rules:
>   - id: no-direct-libyang-import
>     patterns:
>       - pattern-either:
>           - pattern: import yang
>           - pattern: import yang as $X
>           - pattern: from yang import $X
>     message: "Do not import libyang directly. Use sonic_yang / sonic_yang_ext instead."
>     languages: [python]
>     severity: ERROR
>     paths:
>       exclude:
>         - tests/
> ```
> 
> This would fail CI on any PR that reintroduces the dependency. Could be added in this PR or as a follow-up.

Thanks, mostly used this, but also had to add libyang as a pattern since libyang3 uses that import name.

rookie-who reviewed (approved):
Semgrep rule looks good — covers both `yang` and `libyang` imports. Thanks for picking that up.

LGTM.
